### PR TITLE
Optimizing cpp Vectors class to use 2d tensor

### PR DESCRIPTION
**Pull request by Nayef211**
## Description
- Updating Vectors class to use 2d tensor along with a `unordered map<string, Tensor.view()>`
- After conducting performance benchmarks, it was found that this implementation gives the best mix of performance for lookup and construction
- Added functionality for looking up vectors for several tokens using `get_vectors_by_tokens`
- Added tests for the new function

### Existing FastText - Not Jit Mode
**Construction time from file:** 778.5076372753829
**Construction time from cache:** 6.0479023195803165
**Lookup time:** 132.09116122685373

### FastText Experimental
**Construction time:** 670.8132877256721
**Construction time from cache:** 39.3408519141376
#### FastText Experimental - Not Jit Mode
**Lookup time:** 113.68280134536326
#### FastText Experimental - Jit Mode
**Lookup time:** 224.18116508424282

### CODE REVIEW

File: torchtext/csrc/vectors.cpp [hunk 1]
```diff
@@ -2,79 +2,83 @@
...
+            [](const c10::intrusive_ptr<Vectors> &self) -> std::tuple<
+                std::vector<std::string>, torch::Tensor, torch::Tensor> {
+              std::tuple<std::vector<std::string>, torch::Tensor, torch::Tensor>
                   states(self->tokens_, self->vectors_, self->unk_tensor_);
               return states;
             },
-            // __setstate__
-            [](std::tuple<std::vector<std::string>, std::vector<torch::Tensor>,
+            // __getstate__
+            [](std::tuple<std::vector<std::string>, torch::Tensor,
```
Comment (zhangguanheng66) [inline torchtext/csrc/vectors.cpp:81]: Do we want to support a lookup func for a list of tokens?

**Comment by Nayef211** [comment]
> I would be explicit about the speedup of the performance in the summary of PR.

Adding this to the PR description

File: torchtext/csrc/vectors.cpp (hunk 1 above)
Comment (Nayef211) [reply torchtext/csrc/vectors.cpp:81]: Do you think this would be helpful? I can add this functionality in a followup PR?

Comment (Nayef211) [reply torchtext/csrc/vectors.cpp:92]: Done!

File: test/experimental/test_vectors.py [hunk 2]
```diff
@@ -162,49 +177,49 @@ def test_vectors_from_file(self):
...
-            # The first 3 entries in each vector.
-            expected_glove = {
-                'the': [0.27204, -0.06203, -0.1884],
-                'people': [-0.19686, 0.11579, -0.41091],
-            }
-
-            for word in expected_glove.keys():
-                self.assertEqual(vectors_obj[word][:3], expected_glove[word])
-                self.assertEqual(jit_vectors_obj[word][:3], expected_glove[word])
+    # def test_fast_text(self):
```
Comment (zhangguanheng66) [inline test/experimental/test_vectors.py]: Why do we disable those tests?

Comment (Nayef211) [reply test/experimental/test_vectors.py]: Sorry this was just an intermediate commit. This should have been fixed.

File: torchtext/csrc/vectors.cpp (hunk 1 above)
Comment (cpuhrsch) [reply torchtext/csrc/vectors.cpp:92]: Didn't we want to do this in a separate PR? Can you split it out please? The title doesn't indicate it's part of it and there's no design discussion around the name or potential overload mechanisms for existing getitem or maybe even a generic map function.

**Comment by Nayef211** [comment]
> See inline comment

I removed the `get_vectors_by_tokens` function and am adding a new PR for that. Could I get a stamp?

**Pull request merged**